README.md
# P2-One Card Game

## Description

The program is about the card game "OneCard". The program provides a real-game mode for multiple players to play and a demo mode for playing the game automatically. A log file will log the game detail.

## Getting started

1. enter the `./src` directory

2. in a bash shell, input the command: `make`

3. run `./onecard` program with different arguments. For example, `./onecard --rounds=3 -n 4 --decks=2 -c 8`to start the game in real-game mode., and `./onecard --rounds=3 -n 4 --decks=2 -c 8 -a` to start the game in demo mode.

   For CLI argument help, input `./onecard --help` to print the help interface.

4. When the game starts, a player can type the card name to play a card, or input "draw" to draw a card. An example:
    ```
    draw
    Diamond 9 
    ```

## Key Data Structure

A circular double linked list is used for players, and dynamic arrays storing int type data are widely used for cards.

```C
typedef struct _player {
  int number;                 // player number, e.g. 1, 2...
  int score;                  // player score
  int* player_card;           // player card array
  int size_p;                 // size of player card
  struct _player* next;       // pointer to the next player
  struct _player* previous;   // pointer to the previous player
} player;

typedef struct _deck {
  int* stock;     // stock pile array
  int* discard;   // discard pile array
  int size_s;     // size of the stock pile
  int size_d;     // size of the discard pile
  int top;        // top of the stock pile
} deck;
```

The card information(suit and rank) is stored as an integer ranging form 1 to 52. The corresponding transformation is as followed:
```C
// card number->suit and rank
suit = (card - 1) / 13 + 1;
rank = (card - 1) % 13 + 1;
// suit and rank->card number
card=suit * 13 + rank -14;
```
Here suit is 1, 2, 3 or 4, representing Spade, Heart, Diamond and Club respectively. And rank from 1 to 13 means card rank from 2 to A.


## Program Function Structure

The program uses layer programming method to brings much flexibility while also saving much rewriting when adjusting the code and allowing faster debugging in case of problem. The functions in lower layer can't access functions in upper layer.

### Layer 1-data structure: Functions dealing with data structures directly

- new_player: \
Initialize players

- deck_init: \
Initialize decks

- free_player: \
Free memory of players when the game ends

- free_deck: \
Free memory of decks when the game ends

- delete_player_card: \
Delete one card in player's deck

- discard_exchange: \
Delete one card in player's deck, then put it into discard pile

- check_size_p: \
Check whether a player has played all his or her cards

- find_index: \
Find the index of a card in a player's cards

- num2card: \
Convert number to card(string)

- card2num: \
Convert card(string) to number

- reshuffle: \
Put discarded cards into stock pile then shuffle them

- shuffle: \
Shuffle cards in stock pile

- compare: \
Compare function for quick sort algorithm

- next_player: \
Return pointer to the "next" player by clockwise/counterclockwise direction

- show_card_deck: \
Print cards in stock pile

- show_card_deck_log: \
Print cards in stock pile(in log file)

- show_card: \
Print player's cards

- clear_screen: \
Clean the screen

- pause_t: \
  Pause the program


### Layer 2-game: Functions needed to play the game

- deal_card: \
Deal cards to every player

- play_order: \
Decide who is the first player

- auto_detect: \
Help player know what cards can be played

- winner: \
Detect who is the winner

- print_score: \
Get the score and print it out

- calculate_scores: \
Print the leaderboard and winner when the game ends

- sort_players_by_score: \
Sort Players by his or her score

- print_leaderboard: \
Print leaderboard

- find_winners_or_tie: \
Print the result of the game

- reset_player_and_deck: \
Reset the game for next round

- show_all_card: \
Print all players' cards

### Layer 3-interaction: Function needed for the user to interact with the computer and play

- game: \
Organize the game process

- game_play: \
Organize the game process in a round

- play_card: \
play a card

- play_card_demo: \
play a card in demo mod

- draw_card: \
Draw cards

## Game Setting

One Card is a rather simple game played by n persons over a pre-decided number of rounds r . A total of d decks
of Poker cards, excluding Jokers, are shuffled and c cards are offered to each player. Before the game starts each
player is offered an extra card to determine the playing order. This game is then discarded. The game will be played
counter-clockwise, starting with the player who received the extra card with lowest rank. Once the playing order has
been decided all those initial n cards are directly put in the discard pile. Finally, the game starts with the first card of
the stock pile being posed face-up, to initiate the rank and suit.

As the game starts each player, following the defined order, plays exactly one card either following the rank or the suit
defined by the previous card. Any played card directly goes into the discard pile, and anyone who is unable to play
should draw a card from the top of the stock pile. If the stock pile is exhausted, the discard pile is shuffled and used
as stock pile.

As soon as a player has no cards left, the rounds stops. All other players receive a penalty equal to the number of cards
left in their hands. The player who won the round initiates the following one, all the other rules remain unchanged.
At the end of the r rounds the final score of the players is determined by summing up their respective penalties. The
person with highest score wins. In case of equality more than one player can be declared winner.

Cards split into four main categories:
- Attack:
   - Cards with rank 2: the next player draws two cards from the stock pile
   - Cards with rank 3: the next player draws three cards from the stock pile
- Defense:
   - Cards with rank 7: cancel an attack, i.e. do not draw any card if a 2 or a 3 was played before
- Action:
   - Queen cards: reverse the playing order from counter-clockwise to clockwise or clockwise to counterclockwise
   - Jack cards: skip the next player
- Regular: any other card has no special effect and is only used to match the previous card's rank or suit

Notes on cards and attacks:
- The effect of the attack cards is cumulative.
- A Queen or a Jack, of same suit as the previous card, can be played to redirect an attack on the previous player,
or the player after the next one, respectively;
- When attacked, a player not playing a special card (2, 3, 7, Q, J) must draw cards from the stock pile, before
ending its turn;



CHANGELOG.md
# 3.3 (2023-11-19)

**Added**

- Function used in the "real-game" mode to help you know what cards you can play
- Show more information in real-game mode to help players judge

# 3.2 (2023-11-18)

**Added**

- Function to clean screen
- Function to pause the game
- Log file to log game detail

# 3.1 (2023-11-16)

**Added**

- Demo mode
- Leaderboard to better show the final score

# 3.0 (2023-11-13)

**Added**

- Function to deal with special cards and attacked status

**Changed**

- Rewrite the program using a circular double linked list to represent player
- change deck structure
- reorganize functions

# 2.2 (2023-11-5)

**Added**

- Function to calculate the penalty each player has got in the past round, and in the game.
- Print game statistics after each round
- Judge who wins the game
- round setting

# 2.1 (2023-11-4)

**Added**

- Function to decide the first player
- Function to deal cards to players
- Function to let multiple players take turns to play

# 2.0 (2023-11-2)

**Added**

- Initialize the surface
- Function to sort the array
- Function to shuffle the array

# 1.2 (2023-10-29)

**Added**

- Function where a player draws a card
- Function that initializing the game

**Changed**

- The implement logic 

# 1.1 (2023-10-27)

**Added**

- The ability to deal with CLI arguments
- Function showing the current cards in hand
- Function where a player plays a card

# 1.1 (2023-10-26)

**Added**

- Functions to convert cards from number to human language, and from human language to number
- Initializing parameters

# 1.0 (2023-10-24)

**Added**

- Data structure of cards and players

src/game.c
#include "game.h"

// deal cards to every player
void deal_card(player** head, deck** card, int c) {
  player* current = *head;
  do {
    for (int i = 0; i < c; i++) {
      if ((*card)->top >= (*card)->size_s) {
        reshuffle(card, (*card)->size_s + 1);
      }
      current->player_card[i] = (*card)->stock[i + (*card)->top];
    }
    current->size_p = c;
    (*card)->top += c;
    current = current->next;
  } while (current->number != (*head)->number);
}

// decide the first player
int play_order(int** candidate, deck** card, int n, FILE* fd) {
  const int max_attempts = 10;
  int attempt = 0;

  while (attempt < max_attempts) {
    int* canaidate_card = calloc((size_t)n, sizeof(int));
    if (canaidate_card == NULL) {
      printf("Error: out of memory\n");
      exit(0);
    }

    char card_i[20];
    for (int i = 0; i < n; i++) {
      if ((*candidate)[i] == 1) {
        if ((*card)->top >= (*card)->size_s) {
          reshuffle(card, (*card)->size_s + 1);
        }
        canaidate_card[i] = (*card)->stock[i + (*card)->top];
        (*card)->discard[(*card)->size_d] = canaidate_card[i];
        (*card)->top++;
        (*card)->size_d++;
        num2card(canaidate_card[i], card_i);
        printf("Player %d: %s\n", i + 1, card_i);
        fprintf(fd, "Player %d: %s\n", i + 1, card_i);
      } else if ((*candidate)[i] == 0) {
        canaidate_card[i] = 14;
      }
    }
    // select the lowest rank
    for (int i = 0; i < n; i++) {
      for (int j = 0; j < n; j++) {
        if (canaidate_card[i] > canaidate_card[j]) (*candidate)[i] = 0;
      }
    }

    // find the first player
    int count = 0, first = 0;
    for (int i = 0; i < n; i++) {
      if ((*candidate)[i] == 1) {
        count++;
        first = i + 1;
      }
    }
    free(canaidate_card);

    if (count == 1) {
      return first;
    } else {
      printf("\nThere exists same ranks! Another extra card round again.\n\n");
      fprintf(fd, "\nThere exists same ranks! Another extra card round again.\n\n");

      // reset candidate
      for (int i = 0; i < n; i++) {
        (*candidate)[i] = 1;
      }

      attempt++;
    }
  }
  return rand() % n + 1;
}

// help player know what cards can be played
void auto_detect(player** current, const int* current_card, int* attack_sum) {
  if (*current_card == 0) {
    printf("You can play: ");
    char card_i[20];
    for (int i = 0; i < (*current)->size_p; i++) {
      memset(card_i, 0, (size_t)20 * sizeof(char));
      num2card((*current)->player_card[i], card_i);
      printf("%s", card_i);
      if (i < (*current)->size_p - 1) printf(", ");
    }
    printf("\n");
  } else if (*attack_sum == 0) {
    int count = 0;
    printf("You can play: ");
    for (int i = 0; i < (*current)->size_p; i++) {
      int next_card = (*current)->player_card[i];
      int current_card_suit = (*current_card - 1) / 13 + 1;
      int current_card_rank = (*current_card - 1) % 13 + 1;
      int next_card_suit = (next_card - 1) / 13 + 1;
      int next_card_rank = (next_card - 1) % 13 + 1;
      char card_i[20] = {0};
      num2card(next_card, card_i);
      if (next_card_rank == 10 || next_card_rank == 11 || next_card_suit == current_card_suit || next_card_rank == current_card_rank) {
        if (count == 0) {
          printf("%s", card_i);
        } else {
          printf(", %s", card_i);
        }
        count++;
      }
    }
    if (count == 0) printf("None. Type \"draw\" to draw a card.");
  } else {
    int count = 0;
    printf("You can play: ");
    for (int i = 0; i < (*current)->size_p; i++) {
      int next_card = (*current)->player_card[i];
      int next_card_rank = (next_card - 1) % 13 + 1;
      char card_i[20] = {0};
      num2card(next_card, card_i);
      if (next_card_rank == 1 || next_card_rank == 2 || next_card_rank == 6 || next_card_rank == 10 || next_card_rank == 11) {
        if (count == 0) {
          printf("%s", card_i);
        } else {
          printf(", %s", card_i);
        }
        count++;
      }
    }
    if (count == 0) printf("None. Type \"draw\" to draw %d cards.", *attack_sum);
  }

  printf("\n");
}

// detect who is the winner
int winner(player** head) {
  int winner_number = 0;
  player* current = *head;
  while (winner_number == 0) {
    if (current->size_p == 0) winner_number = current->number;
    current = current->next;
  }
  return winner_number;
}

// get the score and print it out
void print_score(player** head, FILE* fd) {
  player* current = *head;
  do {
    printf("Player %d: %d, ", current->number, -current->size_p);
    fprintf(fd, "Player %d: %d,\t ", current->number, -current->size_p);
    current->score -= current->size_p;
    printf("total: %d\n", current->score);
    fprintf(fd, "total: %d\n", current->score);
    current = current->next;
  } while (current != *head);
}

// calcualte the score and print the leaderboard and winner when the game ends
void calculate_scores(player** head, FILE* fd) {
  sort_players_by_score(head);
  print_leaderboard(head, fd);
  find_winners_or_tie(head, fd);
}

// sort players by his or her score
void sort_players_by_score(player** head) {
  if (*head == NULL || (*head)->next == *head) return;

  int swapped = 0;
  do {
    swapped = 0;
    player* current = *head;

    while (current->next != *head) {
      if (current->score < current->next->score) {
        // swap
        int temp_score = current->score;
        current->score = current->next->score;
        current->next->score = temp_score;

        int temp_number = current->number;
        current->number = current->next->number;
        current->next->number = temp_number;

        swapped = 1;
      }
      current = current->next;
    }
  } while (swapped);
}

// print leaderboard
void print_leaderboard(player** head, FILE* fd) {
  printf("\n----Leaderboard----\n");
  fprintf(fd, "\n----Leaderboard----\n");
  player* current = *head;
  int rank = 1;
  do {
    printf("Rank %d: Player %d - Score: %d\n", rank, current->number, current->score);
    fprintf(fd, "Rank %d: Player %d - Score: %d\n", rank, current->number, current->score);
    current = current->next;
    rank++;
  } while (current != *head);
}

// print the result of the game
void find_winners_or_tie(player** head, FILE* fd) {
  int high_score = (*head)->score;
  player* current = (*head)->next;
  int tie = 0;

  // check whether it is a tie
  do {
    if (current->score == high_score && current != *head) {
      tie = 1;
      break;
    }
    current = current->next;
  } while (current != *head);

  if (tie == 1) {
    printf("It's a tie between the following players with a score of %d:\n", high_score);
    fprintf(fd, "It's a tie between the following players with a score of %d:\n", high_score);
    current = *head;
    do {
      if (current->score == high_score) {
        printf("Player %d\n", current->number);
        fprintf(fd, "Player %d\n", current->number);
      }
      current = current->next;
    } while (current != *head);
  } else {
    printf("\nThe winner is Player %d!\n", (*head)->number);
    fprintf(fd, "\nThe winner is Player %d!\n", (*head)->number);
  }
}

// reset the game for next round
void reset_player_and_deck(player** head, deck** card, int size_0) {
  // player card
  player* current = *head;
  do {
    memset(current->player_card, 0, (size_t)current->size_p * sizeof(int));
    current->size_p = 0;
    current = current->next;
  } while (current != *head);

  // discard pile
  memset((*card)->discard, 0, (size_t)(*card)->size_d * sizeof(int));
  (*card)->size_d = 0;

  // stock pile
  int* new_stock = realloc((*card)->stock, (size_t)size_0 * sizeof(int));
  if (new_stock == NULL) {
    printf("Reshuffle memory allocation failed.\n");
    exit(1);
  }
  (*card)->stock = new_stock;
  for (int i = 0; i < size_0; i++) (*card)->stock[i] = 0;
  (*card)->top = 0;
  (*card)->size_s = size_0;
}

// print all players' cards
void show_all_card(player** head, int demo, FILE* fd) {
  player* current = *head;
  do {
    show_card(&current, fd, 1);
    current = current->next;
    if (demo == 0) {
      pause_t();
      clear_screen();
    }
  } while (current->number != (*head)->number);
}


src/main.c
#include "interaction.h"

int main(int argc, char* argv[]) {
  int n = 4, c = 5, d = 2, r = 1;  // n players, deal c cards per player, use d decks, play r rounds
  int demo = 0;                    // demo mode
  int opt = 0, option_index = 0;   // CLI parameter
  char* file = "onecard.log";      // log file name
  struct option long_options[] = {{"help", no_argument, NULL, 'h'},
                                  {"log", required_argument, NULL, 'l'},
                                  {"player-number", required_argument, NULL, 'n'},
                                  {"initial-cards", required_argument, NULL, 'c'},
                                  {"decks", required_argument, NULL, 'd'},
                                  {"rounds", required_argument, NULL, 'r'},
                                  {"auto", no_argument, NULL, 'a'},
                                  {NULL, 0, NULL, 0}};
  while ((opt = getopt_long(argc, argv, ":hn:c:d:r:a", long_options, &option_index)) != -1) {
    switch (opt) {
      case 'h': {
        printf("-h|--help \t\tprint this help message\n");
        printf("--log filename \t\twrite the logs in filename (default: onecard.log)\n");
        printf("-n n|--player-number=n \tn players, n must be larger than 2 (default: 4)\n");
        printf("-c c|--initial-cards=c \tdeal c cards per player, c must be at least 2 (default: 5)\n");
        printf("-d d|--decks=d \t\tuse d decks 52 cards each, d must be at least 2 (default: 2)\n");
        printf("-r r|--rounds=r \tplay r rounds, r must be at least 1 (default: 1)\n");
        printf("-a|--auto \t\trun in demo mode\n");
        return 0;
      }
      case 'l': {
        file = (optarg == NULL ? "onecard.log" : optarg);
        break;
      }
      case 'n': {
        n = atoi(optarg);
        printf("Number of players: %s\n", optarg);
        break;
      }
      case 'c': {
        c = atoi(optarg);
        printf("Number of cards dealt to per player: %s\n", optarg);
        break;
      }
      case 'd': {
        d = atoi(optarg);
        printf("Number of decks: %s\n", optarg);
        break;
      }
      case 'r': {
        r = atoi(optarg);
        printf("Number of rounds: %s\n", optarg);
        break;
      }
      case 'a': {
        demo = 1;  // need to be added
        printf("Run in demo mode\n");
        break;
      }
      case '?': {
        printf("Unknown option %c\n", optopt);
        break;
      }
      case ':': {
        printf("Missing option for %c\n", optopt);
        break;
      }
      default:
        printf("%d %d %d %d %d", demo, n, c, d, r);  // for compiling the file successful
    }
  }
  // detect invalid input
  if (n <= 1 || c <= 0 || d <= 0 || r <= 0) {
    printf("Invalid input.\n");
    return 0;
  }
  if (n * c + 5 >= 52 * d) {
    printf("The number of decks is not sufficient.\n");
    return 0;
  }
  // misc
  srand((unsigned)time(NULL));
  FILE* fd = fopen(file, "w+");
  // start the game
  game(n, c, d, r, demo, fd);
  fclose(fd);
  return 0;
}

// game mode
void game(int n, int c, int d, int r, int demo, FILE* fd) {
  // Initialize the whole game
  printf("########################\n#                      #\n# Welcome to One Card! #\n#                      #\n########################\n\n");
  printf("---- Initial setup ----\nNumber of players: %d\nCards dealt to per player: %d\nNumber of decks: %d\nNumber of rounds: %d\n\n", n, c, d, r);
  fprintf(fd, "########################\n#                      #\n#       GAME LOG       #\n#                      #\n########################\n\n");
  fprintf(fd, "---- Initial setup ----\nNumber of players: \t\t\t%d\nCards dealt to per player: %d\nNumber of decks: \t\t\t%d\nNumber of rounds: \t\t\t%d\n\n", n, c, d, r);
  // Initialize players
  player* head = NULL;
  for (int i = 1; i <= n; i++) new_player(&head, i, d);
  // Initialize decks
  deck* card = calloc((size_t)1, sizeof(deck));
  if (card == NULL) {
    printf("Memory allocation failed.\n");
    exit(1);
  }
  deck_init(&card, d);
  // enter each round
  for (int i = 1; i <= r; i++) {
    printf("-----Round %d-----\n", i);
    fprintf(fd, "-----Round %d-----\n", i);
    if (i > 1) {
      printf("\n\n\n");
      fprintf(fd, "\n\n\n");
      reset_player_and_deck(&head, &card, 52 * d);
    }
    // shuffle cards
    printf("Shuffling cards...\n");
    fprintf(fd, "Shuffling cards...\n");
    for (int j = 0; j < card->size_s; j++) card->stock[j] = (((j + 1) % 52 != 0) ? ((j + 1) % 52) : 52);  // add cards
    shuffle(&card);
    show_card_deck_log(card, fd);
    if (demo == 1) {
      printf("Shuffling results:\n");
      show_card_deck(card);
    }
    // deal cards
    fprintf(fd, "\n\nDealing cards...\n");
    printf("\n\nDealing cards...\n");
    deal_card(&head, &card, c);
    // show cards
    show_all_card(&head, demo, fd);
    // determine play order
    printf("\nDetermining the playing order...\n");
    fprintf(fd, "\nDetermining the playing order...\n");
    int* candidate = calloc((size_t)n, sizeof(int));
    if (candidate == NULL) {
      printf("Memory allocation failed.\n");
      exit(1);
    }
    for (int j = 0; j < n; j++) candidate[j] = 1;
    int first = play_order(&candidate, &card, n, fd);
    printf("The game will start with player %d\n\n", first);
    fprintf(fd, "The game will start with player %d\n\n", first);
    free(candidate);
    candidate = NULL;
    if (demo == 0) pause_t();
    // real game start
    game_play(&head, &card, first, demo, fd);
    // calculate the score
    int winner_number = winner(&head);
    printf("\nPlayer %d wins!\n", winner_number);
    printf("\n---- Stats ----\nRound %d result:\n", i);
    fprintf(fd, "\nPlayer %d wins!\n", winner_number);
    fprintf(fd, "\n---- Stats ----\nRound %d result:\n", i);
    print_score(&head, fd);
    printf("Round %d ends.\n", i);
    fprintf(fd, "Round %d ends.\n\n\n", i);
    if (demo == 0) {
      pause_t();
      clear_screen();
    }
  }
  // final score and winner
  printf("\nAll rounds end.\n");
  fprintf(fd, "\nAll rounds end.\n");
  calculate_scores(&head, fd);
  // free memory: player and deck
  free_player(&head);
  head = NULL;
  free_deck(&card);
  card = NULL;
}


src/data_structure.h
#ifndef DATA_STRUCTURE
#define DATA_STRUCTURE
// Layer 1

#include <getopt.h>
#include <math.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

// data structures
typedef struct _player {
  int number;                // player number, e.g. 1, 2...
  int score;                 // player score
  int* player_card;          // player card array
  int size_p;                // size of player card
  struct _player* next;      // pointer to the next player
  struct _player* previous;  // pointer to the previous player
} player;

typedef struct _deck {
  int* stock;    // stock pile array
  int* discard;  // discard pile array
  int size_s;    // size of the stock pile
  int size_d;    // size of the discard pile
  int top;       // top of the stock pile
} deck;

// functions
void new_player(player** head, int num, int d);
void free_player(player** head);
void free_deck(deck** card);
void delete_player_card(player** current, int player_card_index);
void discard_exchange(player** current, deck** card, int player_card_index, int next_card);
int check_size_p(player** head);
int find_index(player** current, int next_card);

void deck_init(deck** card, int d);
void num2card(int num, char* card_i);
int card2num(char* str);
void reshuffle(deck** card, int size_0);

void shuffle(deck** card);
int compare(const void* a, const void* b);
void next_player(player** current, int is_clockwise);

void show_card_deck(deck* card);
void show_card_deck_log(deck* card, FILE* fd);
void show_card(player** current, FILE* fd, int times);

void clear_screen(void);
void pause_t(void);

#endif


src/data_structure.c
#include "data_structure.h"
// Initialize players
void new_player(player** head, int num, int d) {
  player* newplayer = calloc((size_t)1, sizeof(player));
  if (newplayer == NULL) {
    printf("Error: out of memory\n");
    exit(0);
  }
  int* newcard = calloc((size_t)(d * 52), sizeof(int));
  if (newcard == NULL) {
    printf("Error: out of memory\n");
    exit(0);
  }
  newplayer->player_card = newcard;
  newplayer->size_p = 0;
  newplayer->number = num;
  newplayer->score = 0;
  newplayer->next = NULL;
  newplayer->previous = NULL;
  if (*head == NULL) {
    newplayer->next = newplayer;
    newplayer->previous = newplayer;
    *head = newplayer;
  } else {
    player* tail = (*head)->previous;
    newplayer->next = *head;
    newplayer->previous = tail;
    tail->next = newplayer;
    (*head)->previous = newplayer;
  }
}
// Initialize the deck
void deck_init(deck** card, int d) {
  (*card)->stock = calloc((size_t)(d * 52), sizeof(int));
  if ((*card)->stock == NULL) {
    printf("Memory allocation failed.\n");
    exit(1);
  }
  (*card)->discard = calloc((size_t)(d * 52), sizeof(int));
  if ((*card)->discard == NULL) {
    printf("Memory allocation failed.\n");
    exit(1);
  }
  (*card)->size_s = d * 52;
  (*card)->size_d = 0;
  (*card)->top = 0;
}
// free players
void free_player(player** head) {
  if (*head == NULL) return;

  player* current = *head;
  do {
    player* temp = current;
    current = current->next;
    free(temp->player_card);
    temp->player_card = NULL;
    free(temp);
  } while (current != *head && current != NULL);

  *head = NULL;
}
// free decks
void free_deck(deck** card) {
  if (*card != NULL) {
    if ((*card)->stock != NULL) {
      free((*card)->stock);
      (*card)->stock = NULL;
    }
    if ((*card)->discard != NULL) {
      free((*card)->discard);
      (*card)->discard = NULL;
    }
    free(*card);
    *card = NULL;
  }
}
// shuffle cards
void shuffle(deck** card) {
  for (int i = (*card)->size_s - 1; i > 0; i--) {
    int j = rand() % (i + 1);
    int temp = (*card)->stock[i];
    (*card)->stock[i] = (*card)->stock[j];
    (*card)->stock[j] = temp;
  }
}
// convert number to card
void num2card(int num, char* card_i) {
  if (num < 1 || num > 52) {
    sprintf(card_i, "Invalid Card");
    return;
  }
  const char* suits[] = {"Spade", "Heart", "Diamond", "Club"};
  const char* ranks[] = {"2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A"};
  int suit = (num - 1) / 13;
  int rank = (num - 1) % 13;
  sprintf(card_i, "%s %s", suits[suit], ranks[rank]);
}
// convert card to number
int card2num(char* str) {
  const char* suits[] = {"Spade", "Heart", "Diamond", "Club"};
  const char* ranks[] = {"2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A"};
  char suit[10] = {0}, rank[3] = {0};
  // deal with invalid input str
  if (sscanf(str, "%9s %2s", suit, rank) != 2) {
    return 0;
  }

  int suit_index = -1, rank_index = -1;
  for (int i = 0; i < 4; i++) {
    if (strcmp(suits[i], suit) == 0) {
      suit_index = i;
      break;
    }
  }
  for (int i = 0; i < 13; i++) {
    if (strcmp(ranks[i], rank) == 0) {
      rank_index = i;
      break;
    }
  }
  // deal with invalid input str
  if (suit_index == -1 || rank_index == -1) {
    return 0;
  }

  return suit_index * 13 + rank_index + 1;
}
// quick sort-compare
int compare(const void* a, const void* b) {
  int num1 = *(int*)a;
  int num2 = *(int*)b;

  if (num1 < num2) {
    return -1;
  } else if (num1 > num2) {
    return 1;
  } else {
    return 0;
  }
}
// delete index [n] card in player deck
void delete_player_card(player** current, int player_card_index) {
  for (int i = player_card_index; i < (*current)->size_p - 1; i++) (*current)->player_card[i] = (*current)->player_card[i + 1];
  (*current)->player_card[(*current)->size_p - 1] = 0;
  (*current)->size_p--;
}
// delete index [n] card in player deck, then put it into discard pile
void discard_exchange(player** current, deck** card, int player_card_index, int next_card) {
  (*card)->discard[(*card)->size_d] = next_card;
  (*card)->size_d++;
  delete_player_card(current, player_card_index);
}
// pointer to the next player
void next_player(player** current, int is_clockwise) {
  if (is_clockwise == 1)
    *current = (*current)->next;
  else
    *current = (*current)->previous;
}
// put discarded cards into stock pile then shuffle them
void reshuffle(deck** card, int size_0) {
  (*card)->stock = realloc((*card)->stock, (size_t)size_0 * sizeof(int));
  if ((*card)->stock == NULL) printf("Reshuffle memory allocation failed.\n");

  for (int i = 0; i < size_0; i++) {
    (*card)->stock[i] = (*card)->discard[i];
    (*card)->discard[i] = 0;
  }
  (*card)->size_d = 0;
  for (int i = 0; i < size_0; i++) {
    if ((*card)->stock[i] == 0) {
      (*card)->size_s = i;
      break;
    }
  }
  (*card)->top = 0;
  (*card)->stock = realloc((*card)->stock, (size_t)(*card)->size_s * sizeof(int));
  if ((*card)->stock != NULL) {
    // shuffle cards again
    shuffle(card);
  } else {
    printf("Reshuffle memory allocation failed.\n");
  }
}

// check whether a player has played all his or her cards
int check_size_p(player** head) {
  player* current = *head;
  int end = 0;
  do {
    if (current->size_p == 0) {
      end = 1;
      break;
    }
    current = current->next;
  } while (current->number != 1);
  return end;
}
// find the index of a card
int find_index(player** current, int next_card) {
  int player_card_index = -1;
  for (int i = 0; i < (*current)->size_p; i++) {
    if (next_card == (*current)->player_card[i]) {
      player_card_index = i;
      break;
    }
  }
  return player_card_index;
}
// print player's cards
void show_card(player** current, FILE* fd, int times) {
  if ((*current)->size_p == 0) {
    printf("Player %d have played all cards!", (*current)->number);
    if (times == 1) {
      fprintf(fd, "Player %d have played all cards!", (*current)->number);
    }
  } else {
    qsort((*current)->player_card, (size_t)(*current)->size_p, sizeof(int), compare);  // quick sort
    char card_i[20];
    printf("Player %d cards: ", (*current)->number);
    if (times == 1) {
      fprintf(fd, "Player %d cards: ", (*current)->number);
    }
    for (int i = 0; i < (*current)->size_p; i++) {
      memset(card_i, 0, (size_t)20 * sizeof(char));
      num2card((*current)->player_card[i], card_i);
      printf("%s", card_i);
      if (times == 1) {
        fprintf(fd, "%s", card_i);
      }
      if (i < (*current)->size_p - 1) {
        printf(", ");
      }
      if (times == 1) {
        if (((*current)->player_card[i] >= 27 && (*current)->player_card[i] <= 39) || (*current)->player_card[i] == 9 || (*current)->player_card[i] == 22)
          fprintf(fd, "\t");
        else
          fprintf(fd, "\t\t");
      }
    }
    printf("\n");
    if (times == 1) {
      fprintf(fd, "\n");
    }
  }
}
// print cards in stock pile
void show_card_deck(deck* card) {
  char card_i[20];
  for (int i = 0; i < card->size_s; i++) {
    memset(card_i, 0, (size_t)20 * sizeof(char));
    num2card(card->stock[i], card_i);
    printf("%s", card_i);
    if (i < card->size_s - 1) printf(", ");
  }
}
// print cards in stock pile(in log file)
void show_card_deck_log(deck* card, FILE* fd) {
  fprintf(fd, "Shuffling results:");
  char card_i[20];
  for (int i = 0; i < card->size_s; i++) {
    memset(card_i, 0, (size_t)20 * sizeof(char));
    num2card(card->stock[i], card_i);
    if (i % 10 == 0) fprintf(fd, "\n");
    fprintf(fd, "%s", card_i);
    if ((card->stock[i] >= 27 && card->stock[i] <= 39) || card->stock[i] == 9 || card->stock[i] == 22)
      fprintf(fd, "\t");
    else
      fprintf(fd, "\t\t");
  }
}
// clean the screen
void clear_screen(void) {
#if defined(_WIN32) || defined(_WIN64)
  // Windows system
  int value = system("cls");
#else
  // Unix-like system
  int value = system("clear");
#endif

  if (value != 0) printf("Screen clean failed.\n");
}
// pause the program
void pause_t(void) {
  printf("Press Enter to continue...\n");
  getchar();
}
